### PR TITLE
parser: implement Restore for CreateIndexStmt

### PR DESCRIPTION
**Pull request by firekillice**
Issue: pingcap/tidb#8532

### CODE REVIEW

**Comment by leoppro** [comment]
thanks for your contribution!

File: ast/ddl_test.go [hunk 1]
```diff
@@ -187,3 +187,7 @@ func (ts *testDDLSuite) TestColumnPositionRestore(c *C) {
 	}
 	RunNodeRestoreTest(c, testCases, "alter table t add column a varchar(255) %s", extractNodeFunc)
 }
+
+func (ts *testDDLSuite) TestCreateIndexStmtRestore(c *C) {
```
Comment (leoppro) [inline ast/ddl_test.go]: we test `StmtNode` in `parser_test.go`, like this: #114

the cases of `CREATE INDEX` at `parser_test.go#L1781`

**Comment by firekillice** [comment]
PTAL @leoppro

File: ast/ddl_test.go (hunk 1 above)
Comment (leoppro) [reply ast/ddl_test.go]: remove this function please
rest LGTM

Comment (leoppro) [reply ast/ddl_test.go]: remove this function please
rest LGTM

**Comment by leoppro** [comment]
@kennytm PTAL

**Comment by firekillice** [comment]
PTAL @leoppro

**Comment by leoppro** [comment]
LGTM 
PTAL @kennytm

File: ast/ddl.go [hunk 2]
```diff
@@ -921,7 +921,38 @@ type CreateIndexStmt struct {
...
+		return errors.Annotate(err, "An error occurred while restore CreateIndexStmt.Table")
+	}
+
+	ctx.WritePlain(" (")
+	for i, indexColName := range n.IndexColNames {
+		if i != 0 {
+			ctx.WritePlain(", ")
+		}
+		if err := indexColName.Restore(ctx); err != nil {
+			return errors.Annotate(err, "An error occurred while restore CreateIndexStmt.IndexColNames")
```
Comment (leoppro) [inline ast/ddl.go]: add index into error info

Comment (firekillice) [reply ast/ddl.go]: done

File: ast/ddl.go [hunk 3]
```diff
@@ -921,7 +921,38 @@ type CreateIndexStmt struct {
...
-	return errors.New("Not implemented")
+	ctx.WriteKeyWord("CREATE ")
+	if n.Unique {
+		ctx.WriteKeyWord("UNIQUE ")
+	}
+	ctx.WriteKeyWord("INDEX ")
+	ctx.WriteName(n.IndexName)
+	ctx.WritePlain(" ")
+
+	ctx.WriteKeyWord("ON ")
```
Comment (kennytm) [inline ast/ddl.go]:
```suggestion
	ctx.WriteKeyWord(" ON ")
```

and remove line 930

Comment (firekillice) [reply ast/ddl.go]: done， PTAL ,  thanks

Comment (leoppro) [reply ast/ddl.go]: @kennytm

File: parser_test.go [hunk 4]
```diff
@@ -1789,13 +1789,13 @@ func (s *testParserSuite) TestDDL(c *C) {
...
-		{"CREATE INDEX idx ON t (a) LOCK=NONE", true, ""},
-		{"CREATE INDEX idx USING BTREE ON t (a) USING HASH COMMENT 'foo'", true, ""},
-		{"CREATE INDEX idx USING BTREE ON t (a)", true, ""},
+		{"CREATE INDEX idx ON t (a)", true, "CREATE INDEX `idx` ON `t` (`a`)"},
+		{"CREATE INDEX idx ON t (a) USING HASH", true, "CREATE INDEX `idx` ON `t` (`a`) USING HASH"},
+		{"CREATE INDEX idx ON t (a) COMMENT 'foo'", true, "CREATE INDEX `idx` ON `t` (`a`) COMMENT 'foo'"},
+		{"CREATE INDEX idx ON t (a) USING HASH COMMENT 'foo'", true, "CREATE INDEX `idx` ON `t` (`a`) USING HASH COMMENT 'foo'"},
+		{"CREATE INDEX idx ON t (a) LOCK=NONE", true, "CREATE INDEX `idx` ON `t` (`a`)"},
+		{"CREATE INDEX idx USING BTREE ON t (a) USING HASH COMMENT 'foo'", true, "CREATE INDEX `idx` ON `t` (`a`) USING HASH COMMENT 'foo'"},
+		{"CREATE INDEX idx USING BTREE ON t (a)", true, "CREATE INDEX `idx` ON `t` (`a`) USING BTREE"},
```
Comment (kennytm) [inline parser_test.go]: Can we also add a test case involving `CREATE UNIQUE INDEX`? I assume the parser can already parse it.

Comment (firekillice) [reply parser_test.go]: done PTAL thanks

**Pull request merged**